src/gg.regression.unity.bots/Runtime/Scripts/StateRecorder/InGameObjectFinder.cs

```diff
@@ -234,37 +267,56 @@ private RecordedGameObjectState CreateStateForTransform(int screenWidth, int scr
                         maxY = y;
                     }
 
-                    if (includeOnlyOnCameraObjects
-                        && (
-                            ((minY < 0 || minY > screenHeight - 1)
-                             && (maxY < 0 || maxY > screenHeight - 1)
-                            ) ||
-                            ((minX < 0 || minX > screenWidth - 1)
-                             && (maxX < 0 || maxX > screenWidth - 1)
-                            )
-                        ))
+                    var z = screenSpaceObjectCorner.z;
+                    if (z < minZ)
+                    {
+                        minZ = z;
+                    }
+
+                    if (z > maxZ)
+                    {
+                        maxZ = z;
+                    }
+                }
+
+                if (includeOnlyOnCameraObjects)
+                {
+                    var yBuffer = 0;
+                    var xBuffer = 0;
+                    if (replay)
+                    {
+                        // look a bit outside the view on replay to account for edge cases and small intentional variances in the game code
+                        yBuffer = screenHeight / 2;
+                        xBuffer = screenWidth / 2;
+                    }
+
+                    var xLowerLimit = 0 - xBuffer;
+                    var xUpperLimit = screenWidth + xBuffer;
+                    var yLowerLimit = 0 - yBuffer;
+                    var yUpperLimit = screenWidth + yBuffer;
+                    if (!(minX <= xUpperLimit && maxX >= xLowerLimit && minY <= yUpperLimit && maxY >= yLowerLimit))
                     {
-                        // not in camera.. stop iterating
+                        // not in camera..
                         onCamera = false;
-                        break;
                     }
                 }
 
                 if (onCamera)
                 {
-                    var size = new Vector3(maxX - minX, maxY - minY);
-                    var center = new Vector3(minX + (maxX - minX) / 2, minY + (maxY - minY) / 2);
+                    // make sure the screen space bounds has a non-zero Z size around 0
+                    // we track the true z offset separately for ease of mouse selection on replay
+                    var size = new Vector3(maxX - minX, maxY - minY, 0.1f);
```

Comment:
(non-blocking) Dangerous bookkeeping idea for later speed improvements - we are doing lots of object allocations in this code - new vectors, lists... might be able to recycle the objects later on. But def not this PR, and plus that can get messy.. just brainstorming a bit

Reply:
always some good things to keep in mind looking for performance

These allocations are currently all required as we're building the full state observed for this frame and each game object needs uniquely allocated information until the end of the frame.  Every PR I optimize a bit further, but re-using state objects between frames isn't one I've gotten to you (it's on my list).  As you'll see in the screenshot though, the Unity API calls we have to make actually do most of the allocation overhead.

The biggest areas we die on performance today
1. Serializing to JSON .. I've optimized this a ton along the way, almost to the point of no-where left to optimize short of just writing less data. <- called ONLY ON KEY FRAMES
2. Order of magnitude less than writing JSON, but next biggest thing is indeed around this function <- called EVERY FRAME

![image](https://github.com/Regression-Games/RGUnityBots/assets/112959031/e5c4f71d-b87e-4100-add5-8a8aede79b82)

Reply:
Got it, good to know! I do wonder when it would make sense to switch to some binary format like protobuf, in my past experience it helps a ton with size for sure, and I believe it's gotta be more CPU performant...

Reply:
For our future reference, here is some benchmarking someone did: https://dotnetcoretutorials.com/protobuf-in-c-net-part-4-performance-comparisons/?expand_article=1

Reply:
Find me a cross-platform unity safe protobuf package and I'll give it a try.. I went down that rabbit hole for a few minutes one day recently before going.. nope, not spending that level of dev time yet